src/sentry/event_hashing.py

```diff
@@ -11,13 +11,292 @@
+    type = None
+
+    def get_hash(self):
+        return None
+
+    def _get_metadata_as_dict(self):
+        return {}
+
+    def as_dict(self):
+        rv = {
+            'type': self.type,
+            'hash': self.get_hash(),
+        }
+        rv.update(self._get_metadata_as_dict())
+        return rv
+
+    def __repr__(self):
+        return '<%s %r (%s)>' % (
+            self.__class__.__name__,
+            self.get_hash(),
+            self.type,
+        )
+
+
+class ChecksumVariant(BaseVariant):
+    """A checksum variant returns a single hardcoded hash."""
+    type = 'checksum'
+
+    def __init__(self, hash):
+        self.hash = hash
+
+    def get_hash(self):
+        return self.hash
+
+
+class ComponentVariant(BaseVariant):
+    """A component variant is a variant that produces a hash from the
+    `GroupComponent` it encloses.
+    """
+    type = 'component'
+
+    def __init__(self, component):
+        self.component = component
+
+    def get_hash(self):
+        return self.component.get_hash()
+
+    def _get_metadata_as_dict(self):
+        return {
+            'component': self.component.as_dict(),
+        }
+
+
+class CustomFingerprintVariant(BaseVariant):
+    """A completely custom fingerprint."""
+    type = 'custom-fingerprint'
+
+    def __init__(self, values):
+        self.values = values
+
+    def get_hash(self):
+        return hash_from_values(self.values)
+
+    def _get_metadata_as_dict(self):
+        return {
+            'values': self.values,
+        }
+
+
+class SaltedComponentVariant(BaseVariant):
+    """A salted version of a component."""
+    type = 'salted-component'
+
+    def __init__(self, values, component):
+        self.values = values
+        self.component = component
+
+    def get_hash(self):
+        if not self.component.contributes:
+            return None
+        final_values = []
+        for value in self.values:
+            if value in DEFAULT_FINGERPRINT_VALUES:
+                final_values.extend(self.component.flatten_values())
+            else:
+                final_values.append(value)
+        return hash_from_values(final_values)
+
+    def _get_metadata_as_dict(self):
+        return {
+            'values': self.values,
+            'component': self.component,
+        }
+
+
+def hash_from_values(values):
     result = md5()
-    for bit in hash_bits:
-        result.update(force_bytes(bit, errors='replace'))
+    for value in values:
+        result.update(force_bytes(value, errors='replace'))
     return result.hexdigest()
 
 
+def get_calculated_grouping_variants_for_event(event):
+    """Given an event this returns a dictionary of the matching grouping
+    variants.  Checksum and fingerprinting logic are not handled by this
+    function.  Note that this will completely skip over strategies that
+    were not attempted which is suboptimal from the user experience.
+    """
+    interfaces = event.get_interfaces()
+
+    winning_strategy = None
```

Comment:
Yes. The reason for this is that i'm moving all this code afterwards into a `legacy` strategy and this code will be able to try multiple strategies. So this will later likely report as something like `exception:legacy` or `exception:default`.